qui/tray/disk_space.py

```diff
@@ -159,47 +287,52 @@ def __init__(self, **properties):
         Gtk.main()
 
     def refresh_icon(self):
-        pool_data = PoolUsageData()
-        warning = pool_data.get_warning()
+        pool_data = PoolUsageData(self.qubes_app)
+        vm_data = VMUsageData(self.qubes_app)
+        pool_warning = pool_data.get_warning()
+        vm_warning = vm_data.problematic_vms
 
-        if warning:
+        if pool_warning:
             self.icon.set_from_icon_name("dialog-warning")
             text = _("<b>Qubes Disk Space Monitor</b>\nWARNING! You are "
-                     "running out of disk space.") + ''.join(warning)
+                     "running out of disk space.") + ''.join(pool_warning)
             self.icon.set_tooltip_markup(text)
 
             if not self.warned:
-                notification = Gio.Notification.new(_("Disk usage warning!"))
-                notification.set_priority(Gio.NotificationPriority.HIGH)
-                notification.set_body(
-                    _("You are running out of disk space.") + ''.join(warning))
-                notification.set_icon(
-                    Gio.ThemedIcon.new('dialog-warning'))
-
-                self.send_notification(None, notification)
+                emit_notification(
+                    self,
+                    _("Disk usage warning!"),
+                    _("You are running out of disk space.") + ''.join(
+                        pool_warning))
                 self.warned = True
-
         else:
             self.icon.set_from_icon_name("drive-harddisk")
             self.icon.set_tooltip_markup(
                 _('<b>Qubes Disk Space Monitor</b>\nView free disk space.'))
             self.warned = False
 
+        if vm_warning:
+            for vm_data in vm_warning:
+                vm = vm_data.vm
+                if not vm.features.get('disk-space-not-notify', False):
+                    emit_notification(
+                        self,
+                        _("Qube usage warning"),
+                        _("Qube {} is running out of storage space.".format(
+                            vm.name)),
+                        vm=vm)
```

Comment:
This shows the notification every 2 minutes. You need to save somewhere list of VMs for which you've emitted the warning already.

Reply:
And also, I think it should affect widget's icon.